Lightbulb header accessibility when selected/focused

dpoeschl opened the pull request:
Two major parts of this change
- Porting some old C# to XAML to make theming of the Roslyn part of the lightbulb work. Previously, the top of the lightbulb would bet stuck in whatever theme was first used to show the lightbulb. This was not strictly necessary but very convenient to do, got rid of some bad code, fixed theming of the lightbulb, and helped keep our trigger definitions consistent, making things more maintainable going forward.
- Defining triggers to set the Hyperlink & Run foregrounds to something readable when the parent of the PreviewPane is focused. The actual implementation is somewhat hacky, and would be better if the editor exposed a property that let us know the PreviewPanel has keyboard focus, but it's probably too late / onerous to try to coordinate a mutual fix there. This part fixes the actual bug.
    - This is done in two parts. One is fixing the actual foreground color when focused. We set it to a definitely-contrasting color, which works exactly as desired in High Contrast. Unfortunately, it turns the link Black in the Blue Theme, for example, when we'd probably prefer it stay blue. 
    - To compensate for this and to make it more consistent with the rest of the lightbulb, we now underline the link when it is focused instead of showing the dotted border focus visual (which helps in the Blue Theme as mentioned above, but also helps in HC when you can't tell if you're on the right tab stop).

### Customer scenario

A sight-impaired user may have a difficult time navigating to the lightbulb's "CS0248" (or similar) hyperlink in High Contrast themes.

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems?id=488229&_a=edit&triage=true

### Workarounds, if any

The link is visible before it is selected, so the user can see the link text, then repeatedly hit tab/enter until it navigates (because there are too many tab stops and no indication the link is actually selected).

### Risk

The risk is very low. Nothing else builds on this, and it's a simple change. In the very unlikely event a new xaml binding introduced here fails, it won't crash VS or corrupt state (but a color could be wrong).

### Performance impact

Basically none. This is a simple xaml change that is only relevant when the lightbulb is opened (on the user action) and when the user tabs through the control (per user action). 

### Is this a regression from a previous update?

No.

### Root cause analysis

Accessibility was not taken into account when the Roslyn part of the lightbulb was built.

### How was the bug found?

Work on a related bug uncovered this one. Even though it wasn't filed by accessibility testers or a customer, it is a violation that needs to be fixed.

dpoeschl commented:
I don't immediately know why a commit from @jaredpar is included here.

dpoeschl commented on build/config/PublishData.json:
```diff
@@ -23,7 +23,7 @@
         "dev15.7.x-vs-deps": {
             "nugetKind": "PerBuildPreRelease",
-            "version": "2.7.*",
+            "version": "2.8.*",
```
Ignore this. I'll figure out how to get rid of it.

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml:
```diff
@@ -61,6 +62,36 @@
+            <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.ControlLinkTextBrushKey}}" />
+            <Setter Property="TextDecorations" Value="{x:Null}" />
+            <Setter Property="FocusVisualStyle" Value="{x:Null}" />
+            <Style.Triggers>
```
Some of these may be redundant or useless, but they're a direct port from C# and I didn't want to make even more changes here.

dpoeschl commented on src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml:
```diff
@@ -61,6 +62,36 @@
+                    <Setter Property="TextDecorations" Value="Underline" />
+                </Trigger>
+                <DataTrigger Binding="{Binding Path=ParentElement.IsKeyboardFocusWithin, RelativeSource={RelativeSource Mode=FindAncestor, AncestorType={x:Type pp:PreviewPane}}}" Value="True">
+                    <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.CommandBarTextSelectedBrushKey}}" />
```
Full disclaimer, in the light theme for example, this makes it so these hyperlinks turn black when focused (you have to tab past Preview Changes | Document | Project | Solution to get the focus back up to the top). This is exactly what we want for the High Contrast themes, but sub-optimal for the regular themes. Given the timeline (doing this the right way would require a new ShellColor in the VS repo to be coordinated with this change) and accessibility correctness of this approach, I vote we accept this very minor drawback.

olegtk replied:
Why does it need a new color, there are no suitable existing colors?

olegtk replied:
sounds minor indeed, can you please show a screenshot?

dpoeschl replied:
Sure.

![lightbulblinkcolors](https://user-images.githubusercontent.com/235241/35817238-b12d1666-0a51-11e8-9b68-1d396eca951c.gif)

olegtk replied:
Looks ok to me, although I don't understand why you cannot use the same color as "Preview Changes", "Document" etc links in the bottom panel.

Neme12 replied:
it's a little ugly :(

dpoeschl commented:
Tagging for review: @dotnet/roslyn-ide @matthejo

333fred commented on src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml:
```diff
@@ -77,9 +108,12 @@
                             <Border Name="SeverityIconBorder" Height="16" Width="16" Margin="-3,0,0,-2"/>
                             <TextBlock Name="IdTextBlock" Margin="2,0">
-                                <Hyperlink Name="IdHyperlink" IsEnabled="False" RequestNavigate="LearnMoreHyperlink_RequestNavigate"/>
+                                <Hyperlink Name="IdHyperlink" 
```
Trailing whitespace.

olegtk commented on src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml.cs:
```diff
@@ -84,6 +86,20 @@ internal partial class PreviewPane : UserControl, IDisposable
+        private void PreviewPane_Loaded(object sender, RoutedEventArgs e)
+        {
+            ParentElement = Parent as FrameworkElement;
+        }
```
nit: unsubscribe from Loaded?

dpoeschl commented:
Feedback addressed and screenshot/gif attached. @dotnet/roslyn-ide

jinujoseph commented:
Approved via [Link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/488229)

natidea commented:
Please retarget this PR to dev15.6.x to get this in that release.

jcouv commented:
@dpoeschl It seems you pinged the compiler team for review, but I don't see any compiler changes in this PR. Did you mean some other alias?

dpoeschl commented:
@jcouv -- Hahahaha. I changed the base branch to 15.6.x before I had the correct commits in place, so it pulled in ~120 commits that must have spanned the compiler & interactive parts of the code base. I've removed the requests. :)

dpoeschl commented:
retest ubuntu_16_debug_prtest please
